[B] Add bulk permissions operations.  Adds BUKKIT-426

frymaster opened the pull request:
##### The Issue

Currently PermissionAttachments only allow the changing of one permission at a time, recalculating the Permissible's permission cache from scratch each time.  This can be non trivial when many permissions are registered and many permissions need to be altered.
##### Justification and breakdown

This PR adds three methods enabling bulk changing of permissions. `addPermissions` adds new permissions (or alters existing ones) from a map, `setPermissionsTo` explicitly sets the permissions of the attachment to the supplied map, and `unsetPermissions` removes permissions specified in a list.  As the behaviour of existing methods has not been touched (though two superfluous recalculations have been removed) this is not a breaking change.
##### Relevant PR:

This builds upon @Evenprime's PR #466 (though note that implementation is missing the `this` keyword and so never adds the permissions to the attachment) and also adds the `setPermissionsTo()` method.
##### Potential improvements

It is not possible to have both a <Permission> and a <String> version of these methods with the same name, as the method signatures would be identical (Object).  Like @Evenprime I felt that the String methods were more useful; if both String and Permission versions are desired, I would appreciate guidance on how to name both sets of methods.
##### JIRA Ticket:

BUKKIT-426 - https://bukkit.atlassian.net/browse/BUKKIT-426
##### Testing

Unit tests have been added in org.bukkit.PermissionsTest.java and all pass

Test plugin: https://github.com/frymaster/permissions-test

This plugin exercises the new methods and logs the resulting permissions.  Expected output is as per the comments in the file ( https://github.com/frymaster/permissions-test/blob/master/src/main/java/org/_127001/frymaster/permissionstest/PermissionsTest.java )

Compiled jar file: http://frymaster.127001.org/files/permissionstest-1.0.jar

Output: https://gist.github.com/frymaster/5723247 is as expected

riking commented:
Will this break existing plugins?
Why can't you just build the whole attachment before doing the attaching, if performance is a concern?

Also, please provide a test plugin's binary and source so that we can verify that this works.

frymaster commented:
This should not be a breaking change as new methods are added but old ones have not been removed.

An attachment is already attached to a Permissible when constructed (in fact, you get it **from** the Permissible); while it can (obviously) be removed, there is then no way to reattach it.

Test code will be forthcoming.

EdGruberman commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -93,6 +95,38 @@ public void setPermission(Permission perm, boolean value) {
+     *
+     * @param permissions The permissions and values to be set
+     */
+    public void setPermissionsTo(Map<String, Boolean> permissions) {
```
Consider `resetPermissions`

EdGruberman commented on src/main/java/org/bukkit/permissions/PermissibleBase.java:
```diff
@@ -106,7 +106,6 @@ public PermissionAttachment addAttachment(Plugin plugin, String name, boolean va
         result.setPermission(name, value);
 
-        recalculatePermissions();
 
```
Might as well delete the empty line after also.

frymaster replied:
Oops, thought I had :D

frymaster commented:
Updated PR with test plugin and results

riking commented on src/main/java/org/bukkit/permissions/PermissibleBase.java:
```diff
@@ -106,8 +106,6 @@ public PermissionAttachment addAttachment(Plugin plugin, String name, boolean va
         PermissionAttachment result = addAttachment(plugin);
         result.setPermission(name, value);
 
-        recalculatePermissions();
```
Hmmmmmmmmmmmmmmmmmmmm ignore that, setPermission does a recalc. lol.
**[Deleted other comment because it was a bit big]**

riking commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -93,6 +100,41 @@ public void setPermission(Permission perm, boolean value) {
+     * Adds/changes all the permissions in the supplied map.
+     * <p>
+     * As the {@link Permissible}'s net permissions are only recalculated at the
+     * the end of the operation, this is faster than calling
```
"at the end of the operation" - this sounds weird to say. Also, notes about it should be added to the single methods as well.

Maybe something like this, added to the doc of every method that does recalculation?

``` java
/**
 * Sets a permission to the given value by its fully qualified name.
 * <p>
 * This method will recalculate the permissions on the attached {@link Permissible}. If it is
 * desired to set multiple permissions at once,
 * {@link #setPermissions(blah blah blah blah blah blah)} is preferred, as the recalculation is
 * only done once.
 * 
 * @param name Name of the permission
 * @param value New value of the permission
 * @see setPermissions(Map<String, Boolean> permissions)
 */
public void setPermission(String name, boolean value) {
```

Hmm.. still don't like the language. I'll be thinking.

riking commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -105,6 +147,23 @@ public void unsetPermission(String name) {
+     * @param permissions The permissions to be removed. Non-existent
+     * permissions will be ignored.
+     */
+    public void unsetPermissions(List<String> permissions) {
```
~~I think this method should be moved down one to flow better.
Above, you have the order as (String, Perm, Map), and here you have (String, List, Perm) - they should both be (String, Perm, MULT)~~

frymaster replied:
This reordering has now been done, but the diff chose to interpret it as moving the _other_ methods up, rather than this once down ;)

riking replied:
Well, it looks fine now. Strikethroughing first comment.

riking commented:
Your test plugin doesn't do anything with child permissions; could we see some testing with that?

Also, it reads like a JUnit test..... _suggests looking at src/test/java_

frymaster commented:
JUnit tests added.  Altered method order and reworded javadocs.

riking commented on src/test/java/org/bukkit/PermissionsTest.java:
```diff
@@ -0,0 +1,228 @@
+        assertTrue(p.hasPermission("perm1"));
+        assertTrue(p.hasPermission("parent1"));
+        assertTrue(p.hasPermission("child1"));
+        assertFalse(p.hasPermission("child2"));  //child2 is defined false
```
These tests look nicely done!
One minor complaint: the comment should be one space away from the `//` at all places

riking commented:
I like the tests. :)

frymaster commented:
So, apparently only 2 of the 6 tests were firing.  Fixed, plus also added in some tests which check for the current behaviour where conflicting permissions (ie when a parent specifies one permission for a child, but the child permission has been explicitly set to the opposite) are set to the last one to be added to the attachment.

Commit:
Add new methods for altering permissions in bulk. Adds BUKKIT-426

Remove duplicate permissions checks.
Add some unit tests for permissions and the add/remove methods of PermissionAttachment.

frymaster commented:
(also, commit squashed)

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -1,13 +1,17 @@
 
 /**
- * Holds information about a permission attachment on a {@link Permissible} object
+ * Holds information about a permission attachment on a {@link Permissible}
```
Existing JavaDocs should not be changed. This should be done in the appropriate repository.

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -1,13 +1,17 @@
+ * object
  */
 public class PermissionAttachment {
+
```
Extra blank line that should be removed?

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -34,7 +38,8 @@ public Plugin getPlugin() {
     }
 
     /**
-     * Sets an object to be called for when this attachment is removed from a {@link Permissible}. May be null.
```
https://github.com/Bukkit/Bukkit/pull/879/files#r6378930

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -43,7 +48,8 @@ public void setRemovalCallback(PermissionRemovedExecutor ex) {
     }
 
     /**
-     * Gets the class that was previously set to be called when this attachment was removed from a {@link Permissible}. May be null.
```
https://github.com/Bukkit/Bukkit/pull/879/files#r6378930

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -61,9 +67,11 @@ public Permissible getPermissible() {
     }
 
     /**
-     * Gets a copy of all set permissions and values contained within this attachment.
```
https://github.com/Bukkit/Bukkit/pull/879/files#r6378930 x2

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -73,6 +81,12 @@ public Permissible getPermissible() {
 
     /**
      * Sets a permission to the given value, by its fully qualified name
+     * <p>
```
~~https://github.com/Bukkit/Bukkit/pull/879/files#r6378930~~
Nevermind.

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -84,6 +98,12 @@ public void setPermission(String name, boolean value) {
 
     /**
      * Sets a permission to the given value
+     * <p>
```
~~https://github.com/Bukkit/Bukkit/pull/879/files#r6378930~~
Nevermind.

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -93,9 +113,48 @@ public void setPermission(Permission perm, boolean value) {
+     *
+     * @param permissions The permissions and values to be added or altered
+     */
+    public void setPermissions(Map<String, Boolean> permissions) {
```
Should this not be `addPermissions` or something along those lines? Currently it suggests that the passed map is going to be the ENTIRE new set of permissions for the Permissible

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -93,9 +113,48 @@ public void setPermission(Permission perm, boolean value) {
     }
 
     /**
+     * Adds/changes all the permissions in the supplied map.
```
More detail please :3

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -93,9 +113,48 @@ public void setPermission(Permission perm, boolean value) {
+     *
+     * @param permissions The permissions and values to be set
+     */
+    public void resetPermissions(Map<String, Boolean> permissions) {
```
This (appears) to be better named `setPermissions` (see https://github.com/Bukkit/Bukkit/pull/879/files#r6379007 )

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -93,9 +113,48 @@ public void setPermission(Permission perm, boolean value) {
      * Removes the specified permission from this attachment.
      * <p>
      * If the permission does not exist in this attachment, nothing will happen.
+     * <p>
```
~~https://github.com/Bukkit/Bukkit/pull/879/files#r6378930~~
Nevermind.

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -116,9 +180,26 @@ public void unsetPermission(Permission perm) {
+    /**
      * Removes this attachment from its registered {@link Permissible}
      *
-     * @return true if the permissible was removed successfully, false if it did not exist
```
https://github.com/Bukkit/Bukkit/pull/879/files#r6378930

turt2live commented:
@frymaster If you are still interested in this please ensure it works on the latest code base and that the mentioned concerns are addressed.

frymaster commented:
@turt2live Yup, still works on latest - will revert those formatting changes and rename `setPermissions` to `addPermissions`

re: the method currently called `resetPermissions` - what's your opinion of `setPermissionsTo` ? It's a slightly cumbersome name but imo captures the meaning better.

turt2live commented:
I'd rather "setPermissions" because it is simple and to the point - You are setting the permissions.

Also, that "Add/change" (currently named setPermissions) JavaDoc should change...

(now that I am thinking about it, maybe `modifyPermissions` is better? The JavaDoc would say something like "Modify the permissions for the permissible, adding any permissions not currently found")

EdGruberman commented:
I originally recommended changing `setPermissionsTo` to `resetPermissions` because the word "To" does not distinguish enough what is happening differently.  The word "reset" implies everything is being cleared first.

`setPermission` has been used to date to show modifying an existing or adding a new while leaving the rest alone.  This new approach of clearing then setting needs to be distinguished appropriately.  I believe "reset" captures that new approach well.  Especially when looking at the three possibilities going forward: "set", "unset", "reset".

If it were me designing the original permissions API, I'd have used "put" and "remove" instead of "set" and "unset".  Then your original suggestion might have made more sense, but it was just too similar to the already used "set" to distinguish itself.

turt2live commented:
`reset` can also mean "reset everything to the default values" - Which is not what is happening. That is also why I am suggesting setPermissions (plural) over setPermission (singular).

EdGruberman commented:
@turt2live All these verbs out of context can mean multiple things.  It's important to consider what they mean relative to one another and in particular within the context of the parameters supplied alongside their usages.

In the case of the reset proposal, the method requires a Map of permissions as input.  It would seem unintuitive to assume that the Map supplied would be ignored completely.  On top of that, being set to default does happen on some level, if not entirely.  If you supply an empty Map, permissions will be set to their default values appropriately.  Any permissions not supplied will be reset to default accordingly.

I don't see how your suggestion of only adding a plural indicator to distinguish how the actions being taken are different is useful.  Additionally, that would conflict with an already named `setPermissions` method which does not clear the permissions beforehand.  Then you'd have to address that name being changed also, but it does exactly what the `setPermission` method does but with multiple.  So it already seems very appropriately named.

turt2live commented:
I suppose it could stay as reset, but I don't agree with it. Considering the other methods are (badly, imo) named, reset would be the best option.

frymaster commented:
Apologies for the tardiness in my response; my new job is consuming more of my time than I'd thought.

1) What was the final consensus concerning the naming of what is currently `setPermissions`, `resetPermissions` and `unsetPermissions` ?

2) In regards to more detail in the function description - what's your opinion on

``` /**
     * For every string in the map, set the corresponding permission to the
     * value of the boolean, overwriting if necessary.  Any pre-existing
     * permissions not changed will be preserved.
     * <p>
     * As the {@link Permissible}'s net permissions are only recalculated once,
     * this is faster than calling
     * {@link #setPermission(java.lang.String, boolean)} repeatedly.
     *
     * @param permissions The permissions and values to be added or altered
     */
```

?

EdGruberman commented:
> reset would be the best option.

Your wording sounds a bit cumbersome.  You might consider something like: "existing permissions not in the supplied map will remain unchanged."  But, yes, clarifying what happens to existing permissions is key.

frymaster commented:
Thanks, that's a much better wording

frymaster committed:
Revert out of scope formatting tweaks
Expand description of addPermissions method

frymaster commented:
Still works on latest code base and all tests pass

turt2live commented:
This would need to be updated whenever possible for review to continue :(

frymaster commented:
From a brief view it looks like the only changes since I last looked are to the javadoc.  I'll go over the PR by the weekend at the latest

frymaster committed:
Merge branch 'master' into permissions-changes

Conflicts:
	src/main/java/org/bukkit/permissions/PermissionAttachment.java

Adjusted line breaks on javadoc

asofold commented:
Great :). This will help a lot for permission plugins and features, at least to get rid of any reflection.

frymaster commented:
commit updated, probably >_>

turt2live commented on src/main/java/org/bukkit/permissions/PermissionAttachment.java:
```diff
@@ -1,7 +1,9 @@
+import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import org.bukkit.plugin.Plugin;
```
There should be a gap between import groupings.

turt2live commented on src/test/java/org/bukkit/PermissionsTest.java:
```diff
@@ -0,0 +1,247 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import org.bukkit.permissions.Permissible;
```
Gap between import groupings

turt2live commented on src/test/java/org/bukkit/PermissionsTest.java:
```diff
@@ -0,0 +1,247 @@
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class PermissionsTest {
```
Almost all methods in this class have incorrect spacing

turt2live commented:
I've been made aware that this is already possible with a workaround. I'm not aware of any workaround however (and the source of the information I've gotten thus far appears to be ignoring me...). @frymaster are you aware of any current workaround that would consider this PR "pointless"?

frymaster commented:
The way [privileges](https://github.com/krinsdeath/Privileges) gets round it is, for every player, it creates a new unique Permission object.  It then uses the [getChildren](http://jd.bukkit.org/dev/apidocs/org/bukkit/permissions/Permission.html#getChildren%28%29) method, which returns the Map storing all child permissions, and directly adds all the actual permissions to this map

It's then this composite Permission object that gets added to the PermissionAttachment.  As only one Permission is added, there's only one permission recalculation call.

Whether a workaround makes something _pointless_ is up to the core devs; personally I don't think that's the "natural" way to use the current API. PermissionsBukkit, bPermissions, Essentials and the wiki tutorial all use the "many permissions added to one attachment" system that this PR would enhance. In fact PermissionsBukkit uses reflection to directly manipulate the PermissionAttachment outwith the API, as does Essentials (along with a comment indicating this is a workaround until #466, Evenprime's version of this functionality, is pulled) in order to avoid the multiple recalculations.  Personally, I would think the use of the word "workaround" would indicate something that can be improved
